File: services/callintelligence/src/com/aios/callintelligence/ResilientCommunicationContextBinding.java
package com.aios.callintelligence;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.util.Log;

import com.aios.context.ICommunicationContext;

/** Generation-safe binding for the optional communication-context service. */
final class ResilientCommunicationContextBinding implements AutoCloseable {
    interface Listener {
        void onConnected(ICommunicationContext service);
        void onDisconnected();
    }

    private static final String TAG = "AiosContextBinding";
    private static final String ACTION = "com.aios.context.COMMUNICATION_CONTEXT_SERVICE";
    private static final String PACKAGE = "com.aios.contextintelligence";
    private static final long CONNECT_TIMEOUT_MILLIS = 15_000L;

    private final Context context;
    private final Listener listener;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final ServiceRebindPolicy rebindPolicy = new ServiceRebindPolicy();
    private ContextConnection activeConnection;
    private ICommunicationContext service;
    private boolean binding;
    private boolean connected;
    private boolean closed;

    private final Runnable rebind = () -> {
        if (rebindPolicy.begin()) bindService();
    };

    private final class ContextConnection implements ServiceConnection {
        final Runnable timeout = () -> onConnectionTimedOut(this);

        @Override
        public void onServiceConnected(ComponentName name, IBinder binder) {
            ICommunicationContext candidate = ICommunicationContext.Stub.asInterface(binder);
            if (candidate == null) {
                replaceTerminalBinding(this, false);
                return;
            }
            synchronized (ResilientCommunicationContextBinding.this) {
                if (closed || activeConnection != this) return;
                service = candidate;
                connected = true;
            }
            mainHandler.removeCallbacks(timeout);
            rebindPolicy.connected();
            notifyConnected(candidate);
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            // Android normally retains a crash binding. Replace it if reconnect stalls.
            if (!clearCurrent(this)) return;
            notifyDisconnected();
            armConnectionTimeout(this);
        }

        @Override
        public void onBindingDied(ComponentName name) {
            replaceTerminalBinding(this, true);
        }

        @Override
        public void onNullBinding(ComponentName name) {
            replaceTerminalBinding(this, false);
        }
    }

    ResilientCommunicationContextBinding(Context context, Listener listener) {
        this.context = context.getApplicationContext();
        this.listener = listener;
    }

    void start() {
        synchronized (this) {
            if (closed || activeConnection != null || binding) return;
        }
        scheduleRebind(true);
    }

    synchronized boolean isCurrent(ICommunicationContext candidate) {
        return !closed && connected && service == candidate;
    }

    void invalidate(ICommunicationContext candidate) {
        ContextConnection connection;
        synchronized (this) {
            if (closed || service != candidate) return;
            connection = activeConnection;
        }
        replaceTerminalBinding(connection, false);
    }

    @Override
    public void close() {
        ContextConnection connection;
        boolean notify;
        synchronized (this) {
            if (closed) return;
            closed = true;
            notify = service != null;
            service = null;
            connected = false;
            connection = activeConnection;
            activeConnection = null;
            binding = false;
        }
        rebindPolicy.close();
        mainHandler.removeCallbacks(rebind);
        if (connection != null) mainHandler.removeCallbacks(connection.timeout);
        if (notify) notifyDisconnected();
        unbindQuietly(connection);
    }

    private void bindService() {
        ContextConnection connection = new ContextConnection();
        synchronized (this) {
            if (closed || activeConnection != null || binding) return;
            activeConnection = connection;
            binding = true;
        }
        boolean didBind = false;
        try {
            Intent intent = new Intent(ACTION).setPackage(PACKAGE);
            didBind = context.bindService(intent, connection, Context.BIND_AUTO_CREATE);
        } catch (RuntimeException error) {
            Log.e(TAG, "Communication Context bind failed", error);
        }

        boolean release = false;
        boolean retry = false;
        synchronized (this) {
            if (activeConnection != connection) {
                release = didBind;
            } else {
                binding = false;
                if (closed) {
                    activeConnection = null;
                    release = didBind;
                } else if (!didBind) {
                    activeConnection = null;
                    retry = true;
                }
            }
        }
        if (release) {
            unbindQuietly(connection);
        } else if (retry) {
            scheduleRebind(false);
        } else {
            armConnectionTimeout(connection);
        }
    }

    private boolean clearCurrent(ContextConnection connection) {
        synchronized (this) {
            if (closed || activeConnection != connection) return false;
            service = null;
            connected = false;
            return true;
        }
    }

    private void replaceTerminalBinding(ContextConnection connection, boolean immediate) {
        if (connection == null) return;
        boolean notify;
        synchronized (this) {
            if (closed || activeConnection != connection) return;
            notify = service != null;
            service = null;
            connected = false;
            activeConnection = null;
            binding = false;
        }
        mainHandler.removeCallbacks(connection.timeout);
        if (notify) notifyDisconnected();
        unbindQuietly(connection);
        scheduleRebind(immediate);
    }

    private void armConnectionTimeout(ContextConnection connection) {
        synchronized (this) {
            if (closed || activeConnection != connection || connected) return;
        }
        mainHandler.removeCallbacks(connection.timeout);
        mainHandler.postDelayed(connection.timeout, CONNECT_TIMEOUT_MILLIS);
    }

    private void onConnectionTimedOut(ContextConnection connection) {
        synchronized (this) {
            if (closed || activeConnection != connection || connected) return;
        }
        replaceTerminalBinding(connection, false);
    }

    private void scheduleRebind(boolean immediate) {
        long delay = rebindPolicy.reserve(immediate);
        if (delay != ServiceRebindPolicy.NO_RETRY) {
            mainHandler.postDelayed(rebind, delay);
        }
    }

    private void notifyConnected(ICommunicationContext candidate) {
        try {
            listener.onConnected(candidate);
        } catch (RuntimeException error) {
            Log.e(TAG, "Communication Context connection listener failed", error);
            invalidate(candidate);
        }
    }

    private void notifyDisconnected() {
        try {
            listener.onDisconnected();
        } catch (RuntimeException error) {
            Log.e(TAG, "Communication Context disconnection listener failed", error);
        }
    }

    private void unbindQuietly(ContextConnection connection) {
        if (connection == null) return;
        try {
            context.unbindService(connection);
        } catch (IllegalArgumentException ignored) {
            // A close or replacement can race bindService completion.
        }
    }
}

File: apps/messaging/src/com/aios/messaging/context/ResilientCommunicationContextBinding.kt
package com.aios.messaging.context

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import com.aios.context.ICommunicationContext

/** Restartable, generation-safe binding for Messaging's Communication Context client. */
internal class ResilientCommunicationContextBinding(
    context: Context,
    private val listener: Listener,
) {
    interface Listener {
        fun onConnected(service: ICommunicationContext)
        fun onDisconnected(service: ICommunicationContext)
    }

    private val application = context.applicationContext
    private val main = Handler(Looper.getMainLooper())
    private var rebindPolicy = MessagingServiceRebindPolicy()
    private var activeConnection: ContextConnection? = null
    private var rebindTask: Runnable? = null
    private var watchdog: Runnable? = null
    private var remote: ICommunicationContext? = null
    private var generation = 0L
    private var ready = false
    private var started = false

    private inner class ContextConnection(val generation: Long) : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, binder: IBinder?) {
            main.post { acceptConnected(this, binder) }
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            main.post { acceptDisconnected(this) }
        }

        override fun onBindingDied(name: ComponentName?) {
            main.post { terminate(this, expected = null, immediate = true) }
        }

        override fun onNullBinding(name: ComponentName?) {
            main.post { terminate(this, expected = null, immediate = false) }
        }
    }

    fun start() {
        check(Looper.myLooper() == Looper.getMainLooper())
        if (started) return
        started = true
        rebindPolicy = MessagingServiceRebindPolicy()
        scheduleRebind(immediate = true)
    }

    fun stop() {
        check(Looper.myLooper() == Looper.getMainLooper())
        if (!started) return
        started = false
        rebindTask?.let(main::removeCallbacks)
        rebindTask = null
        cancelWatchdog()
        rebindPolicy.close()
        val disconnected = remote
        remote = null
        ready = false
        val connection = activeConnection
        activeConnection = null
        disconnected?.let(listener::onDisconnected)
        connection?.let { runCatching { application.unbindService(it) } }
    }

    fun invalidate(expected: ICommunicationContext) {
        main.post {
            val connection = activeConnection ?: return@post
            terminate(connection, expected, immediate = false)
        }
    }

    private fun acceptConnected(connection: ContextConnection, binder: IBinder?) {
        if (!isCurrentGeneration(connection)) return
        val service = ICommunicationContext.Stub.asInterface(binder)
        if (service == null) {
            terminate(connection, expected = null, immediate = false)
            return
        }
        val displaced = remote
        remote = service
        ready = true
        cancelWatchdog()
        rebindPolicy.connected()
        if (displaced != null && displaced !== service) listener.onDisconnected(displaced)
        listener.onConnected(service)
    }

    private fun acceptDisconnected(connection: ContextConnection) {
        if (!isCurrentGeneration(connection)) return
        val disconnected = remote
        remote = null
        ready = false
        disconnected?.let(listener::onDisconnected)
        // Android retains an ordinary crash binding; replace it if reconnect stalls.
        armWatchdog(connection)
    }

    private fun terminate(
        connection: ContextConnection,
        expected: ICommunicationContext?,
        immediate: Boolean,
    ) {
        if (!isCurrentGeneration(connection) ||
            (expected != null && remote !== expected)
        ) return
        cancelWatchdog()
        val disconnected = remote
        remote = null
        ready = false
        activeConnection = null
        disconnected?.let(listener::onDisconnected)
        runCatching { application.unbindService(connection) }
        scheduleRebind(immediate)
    }

    private fun scheduleRebind(immediate: Boolean) {
        if (!started || activeConnection != null) return
        val delay = rebindPolicy.reserve(immediate)
        if (delay == MessagingServiceRebindPolicy.NO_RETRY) return
        lateinit var task: Runnable
        task = Runnable {
            if (rebindTask !== task) return@Runnable
            rebindTask = null
            if (rebindPolicy.begin()) bindNow()
        }
        rebindTask = task
        if (!main.postDelayed(task, delay)) {
            rebindTask = null
            rebindPolicy.close()
        }
    }

    private fun bindNow() {
        if (!started || activeConnection != null) return
        if (generation == Long.MAX_VALUE) return
        val connection = ContextConnection(++generation)
        activeConnection = connection
        ready = false
        val intent = Intent(ACTION).setComponent(ComponentName(SERVICE_PACKAGE, SERVICE_CLASS))
        val bound = runCatching {
            application.bindService(intent, connection, Context.BIND_AUTO_CREATE)
        }.getOrDefault(false)
        if (!bound) {
            activeConnection = null
            scheduleRebind(immediate = false)
            return
        }
        armWatchdog(connection)
    }

    private fun armWatchdog(connection: ContextConnection) {
        cancelWatchdog()
        if (!isCurrentGeneration(connection) || ready) return
        lateinit var task: Runnable
        task = Runnable {
            if (watchdog !== task) return@Runnable
            watchdog = null
            if (isCurrentGeneration(connection) && !ready) {
                terminate(connection, expected = null, immediate = false)
            }
        }
        watchdog = task
        if (!main.postDelayed(task, CONNECT_TIMEOUT_MILLIS)) {
            watchdog = null
            terminate(connection, expected = null, immediate = false)
        }
    }

    private fun cancelWatchdog() {
        watchdog?.let(main::removeCallbacks)
        watchdog = null
    }

    private fun isCurrentGeneration(connection: ContextConnection): Boolean =
        started && activeConnection === connection && connection.generation == generation

    private companion object {
        const val ACTION = "com.aios.context.COMMUNICATION_CONTEXT_SERVICE"
        const val SERVICE_PACKAGE = "com.aios.contextintelligence"
        const val SERVICE_CLASS =
            "com.aios.contextintelligence.CommunicationContextService"
        const val CONNECT_TIMEOUT_MILLIS = 15_000L
    }
}
